File: RNBridgeOC/RNBridgeOC/RNManagerVC.m
//
//  RNManagerVC.m
//  RNBridgeOC
//
//  Created by Zero on 2017/3/27.
//  Copyright © 2017年 macbook. All rights reserved.
//

#import "RNManagerVC.h"
#import "NotificationMacro.h"

@interface RNManagerVC ()
{
    RNBackTypeBlock backBlock_;
    RNNextTypeBlock nextBlock_;
}
@end

@implementation RNManagerVC

- (void)viewDidLoad {
    [super viewDidLoad];
}
/// RNBackTypeBlock
    // RN控制器跳回上级控制器操作
    // typedef void(^RNBackTypeBlock)(id rnParams);
/// RNNextTypeBlock
    // RN控制器跳转别的控制器操作
    // typedef void(^RNNextTypeBlock)(UIViewController *vc, id rnParams);
- (instancetype)initWithBack:(RNBackTypeBlock)rnBackTyp Next:(RNNextTypeBlock)rnNextType {
    if(self = [super init]){
        
        backBlock_ = rnBackTyp;
        nextBlock_ = rnNextType;
    }
    
    return self;
}

/// 不能写在viewDidLoad  和 dealloc 因为有可能有原生跳RN  再跳原生  再跳RN  最后back，发通知会收到两个通知方法
- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [Notif addObserver:self selector:@selector(back:) name:kNotificationRNBackName object:nil];
    [Notif addObserver:self selector:@selector(nextVC:) name:kNotificationRNNextVCName object:nil];
}
- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    
    NotificationAllRemoveFrom(self);
}
- (void)dealloc {
    nextBlock_ = nil;
    backBlock_ = nil;
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.

}

- (void)back:(NSNotification *)notification {
    NSLog(@"current NSThread %@",[NSThread currentThread]);
    
    id params = [notification object];
    
    // 必须在主线程跳转VC, 因为RN会开一条特殊的线程处理
    dispatch_async(dispatch_get_main_queue(), ^{
        
        if (backBlock_) {
            backBlock_(params);
        }
        
        if(self.navigationController.viewControllers.count > 1)
        {
            [self.navigationController popViewControllerAnimated:YES];
        }
        else
        {
            [self dismissViewControllerAnimated:YES completion:nil];
        }
    });
}

- (void)nextVC:(NSNotification *)notification {
    NSLog(@"current NSThread %@",[NSThread currentThread]);
    
    id params = [notification object];
    
    // 必须在主线程跳转VC, 因为RN会开一条特殊的线程处理
    dispatch_async(dispatch_get_main_queue(), ^{
        
        if (nextBlock_) {
            nextBlock_(self, params);
        }
    });
    
}

@end

File: RNBridgeSwift/RNBridgeSwift/RNManagerVC.swift
//
//  RNManagerVC.swift
//  RNBridgeSwift
//
//  Created by Zero on 2017/3/28.
//  Copyright © 2017年 macbook. All rights reserved.
//

import UIKit

typealias BackBlock = (_ message: Any) -> ()

typealias NextBlock = (_ vc: UIViewController, _ message: Any) -> ()

class RNManagerVC: UIViewController {

    var backBlock: BackBlock?
    var nextBlock: NextBlock?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    convenience init(backBlock: @escaping BackBlock, nextBlock: @escaping NextBlock) {
        
        self.init()
        
        self.backBlock = backBlock;
        self.nextBlock = nextBlock;
    }
    
    override func viewWillAppear(_ animated: Bool) {
        NotificationCenter.default.addObserver(self, selector: #selector(RNManagerVC.backPreviousVC(_:)), name: NSNotification.Name.init(rawValue: "NotificatioinBack"), object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(RNManagerVC.backPreviousVC(_:)), name: NSNotification.Name.init(rawValue: "NotificatioinNext"), object: nil)
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        
        NotificationCenter.default.removeObserver(self);
    }
}


extension RNManagerVC {
    func backPreviousVC(_ notification: Notification) {
        
        print("current Thread %@",Thread.current)

        DispatchQueue.main.async { 
            
            if (self.backBlock != nil) {
                self.backBlock!(notification.object ?? "");
            }
            
            if let navigationVC: UINavigationController = self.navigationController, navigationVC.viewControllers.count > 1 {
                
                navigationVC.popViewController(animated: true)
            } else {
                self.dismiss(animated: true, completion: nil)
            }
            
        }
    }
    
    func nextHandle(_ notification: Notification) {
        print("current Thread %@",Thread.current)
 
        DispatchQueue.main.async {
            if (self.backBlock != nil) {
                self.backBlock!(notification.object ?? "");
            }
        }
    }
}
